File: frontend/src/api/endpoints/lens.ts
/*
 * (C) Copyright 2026- ECMWF and individual contributors.
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 * In applying this licence, ECMWF does not waive the privileges and immunities
 * granted to it by virtue of its status as an intergovernmental organisation nor
 * does it submit to any jurisdiction.
 */

/**
 * Lens API client. The lens routes are passthroughs to the backend lens
 * manager, which spawns external tools (e.g. SkinnyWMS) on dynamic ports.
 * Always check status until `running` before talking to the lens directly.
 */

import { z } from 'zod'
import type {
  LensInstanceDetailResponse,
  SupportedLensDetail,
} from '@/api/types/lens.types'
import {
  LensInstanceDetailResponseSchema,
  SupportedLensDetailSchema,
} from '@/api/types/lens.types'
import { apiClient } from '@/api/client'
import { API_ENDPOINTS } from '@/api/endpoints'
import { getBackendBaseUrl } from '@/utils/env'

export async function startSkinnyWms(localPath: string): Promise<string> {
  return apiClient.post(
    API_ENDPOINTS.lens.startSkinnyWms,
    {},
    {
      params: { local_path: localPath },
      schema: z.string(),
    },
  )
}

export async function getLensStatus(
  lensInstanceId: string,
): Promise<LensInstanceDetailResponse> {
  return apiClient.get(API_ENDPOINTS.lens.status, {
    params: { lens_instance_id: lensInstanceId },
    schema: LensInstanceDetailResponseSchema,
  })
}

export async function stopLens(lensInstanceId: string): Promise<string> {
  return apiClient.delete(API_ENDPOINTS.lens.stop, {
    params: { lens_instance_id: lensInstanceId },
    schema: z.string(),
  })
}

export async function listLenses(): Promise<Array<LensInstanceDetailResponse>> {
  return apiClient.get(API_ENDPOINTS.lens.list, {
    schema: z.array(LensInstanceDetailResponseSchema),
  })
}

export async function listSupportedLenses(): Promise<
  Array<SupportedLensDetail>
> {
  return apiClient.get(API_ENDPOINTS.lens.supported, {
    schema: z.array(SupportedLensDetailSchema),
  })
}

/**
 * Build the absolute base URL of a running lens instance from its port.
 * Lens binds on the same host as the FIAB backend; falls through to
 * `window.location` for same-origin deployments.
 */
export function buildLensBaseUrl(port: number): string {
  const backendBase = getBackendBaseUrl()
  const source = backendBase
    ? new URL(backendBase)
    : new URL(window.location.href)
  return `${source.protocol}//${source.hostname}:${port}`
}

/**
 * GetCapabilities URL for the WMS instance behind a lens. Suitable for
 * pasting into external WMS clients (QGIS, ArcGIS, etc.).
 */
export function buildWmsCapabilitiesUrl(port: number): string {
  return `${buildLensBaseUrl(port)}/wms?service=WMS&version=1.3.0&request=GetCapabilities`
}

File: backend/src/forecastbox/routes/lens.py
# (C) Copyright 2024- ECMWF.
#
# This software is licensed under the terms of the Apache Licence Version 2.0
# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
#
# In applying this licence, ECMWF does not waive the privileges and immunities
# granted to it by virtue of its status as an intergovernmental organisation
# nor does it submit to any jurisdiction.

"""
Lens routes — /lens/*. Corresponds to the `domain.lens` domain.

Lenses are external inspection tools (e.g. skinnyWMS) that clients can launch
against Run outputs. Routes cover: start, status, stop, list, and supported lenses.
"""

# TODO currently no authentication here. Add auth and propagate into the manager itself
# TODO currently no log propagation -- consider routing stdout of skinnywms to files, and allow retrieval here via a new route

import logging
import pathlib
from typing import Any, Self

import httpx
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import Response

from forecastbox.domain.lens.manager import (
    LensInstanceDetail,
    LensInstanceId,
    LensInstanceManager,
    LensStatus,
    get_status,
    list_instances,
    start_skinny_wms,
    stop_instance,
)
from forecastbox.utility.concurrent import NoFreePortsException
from forecastbox.utility.pydantic import FiabBaseModel

PREFIX = "/api/v1/lens"

logger = logging.getLogger(__name__)

router = APIRouter(
    tags=["lens"],
    responses={404: {"description": "Not found"}},
)


class LensInstanceDetailResponse(FiabBaseModel):
    """API response model for a lens instance detail. Mirrors LensInstanceDetail fields
    to decouple the public contract from internal domain refactoring."""

    lens_instance_id: LensInstanceId
    status: LensStatus
    lens_name: str
    lens_params: dict[str, Any]
    ports: list[int]

    @classmethod
    def from_detail(cls, lens_instance_id: LensInstanceId, detail: LensInstanceDetail) -> Self:
        return cls(
            lens_instance_id=lens_instance_id,
            status=detail.status,
            lens_name=detail.lens_name,
            lens_params=detail.lens_params,
            ports=list(detail.ports),
        )


class SupportedLensDetail(FiabBaseModel):
    name: str
    route: str
    params: dict[str, str]


@router.post("/start/skinnyWMS")
def start_skinny_wms_endpoint(local_path: str) -> LensInstanceId:
    """Start a skinnyWMS lens instance serving data from the given local path."""
    try:
        if not pathlib.Path(local_path).exists():
            raise HTTPException(status_code=400, detail="Provided path does not exist")
        return start_skinny_wms(local_path)
    except NoFreePortsException:
        raise HTTPException(status_code=503, detail="No free ports available for a new lens instance")
    except TimeoutError:
        raise HTTPException(status_code=503, detail="Lens manager is busy")


@router.get("/status")
def get_lens_status(lens_instance_id: LensInstanceId) -> LensInstanceDetailResponse:
    """Get the status of a lens instance."""
    try:
        return LensInstanceDetailResponse.from_detail(lens_instance_id, get_status(lens_instance_id))
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Lens instance {lens_instance_id!r} not found")


@router.delete("/stop")
def stop_lens(lens_instance_id: LensInstanceId) -> str:
    """Stop and remove a lens instance."""
    try:
        stop_instance(lens_instance_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Lens instance {lens_instance_id!r} not found")
    except TimeoutError:
        raise HTTPException(status_code=503, detail="Lens manager is busy")
    return "ok"


@router.get("/list")
def list_lenses() -> list[LensInstanceDetailResponse]:
    """List all active lens instances with their current status."""
    return [LensInstanceDetailResponse.from_detail(iid, detail) for iid, detail in list_instances()]


# A SkinnyWMS lens binds gunicorn to 127.0.0.1:<port> on the backend pod, which
# is unreachable from the browser. Browsers reach the backend at port 8000
# only, so proxy WMS traffic through the backend's own host. The port must
# belong to an active lens to avoid being used as a generic localhost-port
# probe.
@router.api_route("/proxy/{port}/{path:path}", methods=["GET", "HEAD"])
async def lens_proxy(port: int, path: str, request: Request) -> Response:
    known_ports = {p for _, inst in LensInstanceManager.instances.items() for p in inst.ports}
    if port not in known_ports:
        raise HTTPException(status_code=404, detail=f"Port {port} does not belong to any active lens")
    upstream = f"http://127.0.0.1:{port}/{path}"
    if request.url.query:
        upstream = f"{upstream}?{request.url.query}"
    async with httpx.AsyncClient(timeout=60.0) as client:
        try:
            resp = await client.request(request.method, upstream)
        except httpx.RequestError as exc:
            raise HTTPException(status_code=502, detail=f"Lens upstream error: {exc!r}")
    excluded = {"content-encoding", "content-length", "transfer-encoding", "connection"}
    headers = {k: v for k, v in resp.headers.items() if k.lower() not in excluded}
    return Response(content=resp.content, status_code=resp.status_code, headers=headers, media_type=resp.headers.get("content-type"))


@router.get("/supported")
def list_supported_lenses() -> list[SupportedLensDetail]:
    """List all supported lens types with their start route and parameters."""
    return [
        SupportedLensDetail(
            name="skinnyWMS",
            route=f"{PREFIX}/start/skinnyWMS",
            params={"local_path": "Absolute path to the data directory or file to serve"},
        )
    ]
